core/src/main/java/org/apache/iceberg/FastAppend.java

```diff
@@ -43,6 +44,7 @@ class FastAppend extends SnapshotProducer<AppendFiles> implements AppendFiles {
   private final PartitionSpec spec;
   private final SnapshotSummary.Builder summaryBuilder = SnapshotSummary.builder();
   private final List<DataFile> newFiles = Lists.newArrayList();
+  private final CharSequenceSet newFilePaths = CharSequenceSet.empty();
```

Comment:
subclasses of `DataFile` don't implement `equals()` / `hashCode()`, so using a separate set to store the file path

Reply:
I feel like it might make this cleaner to just implement equal and hash code then use a `LinkedHashSet` instead of the list.

Reply:
Seems like there are issues with implementing hashCode/equals in that there may be different contexts where we only care about the path vs. the entire object representation.